fix(normalize): Allow override context.campaign

SpainTrain opened the pull request:
- If an override context.campaign is provided, do not overwrite with UTM
  from querystring

closes https://github.com/segmentio/analytics.js-core/issues/34

f2prateek commented:
LGTM. cc @stephenmathieson https://github.com/segmentio/analytics.js-core/issues/34#issuecomment-284589145

f2prateek commented on lib/index.js:
```diff
@@ -213,7 +213,9 @@ Segment.prototype.normalize = function(msg) {
   }
-  if (query) ctx.campaign = utm(query);
+  if (query && !ctx.campaign) {
+    ctx.campaign = utm(query);
```
A possible improvement might be to do a merge if `ctx.campaign` exists. e.g. maybe you track referrer yourself, but still want the other context.campaign fields to be filled by us.

SpainTrain replied:
I started to do that initially, but then thought that campaign properties usually go together as a group, so there is some risk of inadvertently merging campaign properties from two different groups (if overrides and UTM differ).

happy to do it either way and I can make the change quickly if desired.

f2prateek replied:
Let's keep it as is for now.

SpainTrain replied:
:+1:

SpainTrain commented:
hmm, first CI run passed, then the second (fixed tpyo in commit message, no code change) failed due to failing assertions about cookies in before/after hooks.

SpainTrain commented:
> hmm, first CI run passed, then the second (fixed tpyo in commit message, no code change) failed due to failing assertions about cookies in before/after hooks.

disregard, kicking it off again made it pass.

stephenmathieson commented on lib/index.js:
```diff
@@ -213,7 +213,9 @@ Segment.prototype.normalize = function(msg) {
     }
   }
-  if (query) ctx.campaign = utm(query);
+  if (query && !ctx.campaign) {
```
I'd like to see a comment here explaining why we're doing this, as it isn't clear without digging

f2prateek replied:
We don't want to override campaign if a user is providing one. This is fairly standard in most of our client libraries, including this one, no?

e.g.

1. https://github.com/SpainTrain/analytics.js-integration-segmentio/blob/bff98739474c1ce92809601cc4f76bb4acc6fad5/lib/index.js#L210
2. https://github.com/SpainTrain/analytics.js-integration-segmentio/blob/bff98739474c1ce92809601cc4f76bb4acc6fad5/lib/index.js#L207

SpainTrain replied:
Added the comment, but can remove based on what you all decide

f2prateek replied:
Thanks —  adding a comment certainly doesn't hurt, just wasn't sure if it was required! No need to remove it now :)

Commit:
fix(normalize): Allow override context.campaign

- If an override context.campaign is provided, do not overwrite with UTM
  from querystring

closes https://github.com/segmentio/analytics.js-core/issues/34

SpainTrain commented:
@stephenmathieson anything else I should fix here? Thanks for the review!

stephenmathieson commented:
Nope, LGTM!

SpainTrain commented:
Awesome.

You guys good to merge and ship?

stephenmathieson commented:
@f2prateek I'm not sure what the process for releasing changes is for this integration. Any way you could handle that?

f2prateek commented:
@stephenmathieson Sure, for future reference it is https://paper.dropbox.com/doc/Analytics.js-Deployment-ZFzuZSXuTQC.

stephenmathieson commented:
Ah ok, thanks dude!

SpainTrain commented:
:tada:

SpainTrain commented:
What is the expected latency before these changes will be in the production asset?  Not trying to be pushy, just want to know when I should test my custom `campaign` tracking code again.  Thanks!

f2prateek commented:
@SpainTrain generally takes about a couple of hours, but I'm going to sneak in a few other changes that have been pending for a while as well though; should be no later than Monday (want to avoid deploying on Friday).
